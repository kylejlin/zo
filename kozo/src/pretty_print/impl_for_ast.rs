use super::*;

use crate::syntax_tree::ast::*;

impl Display for PrettyPrinted<'_, Expr> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_expr(self.0, f, Indentation { soft_tab_count: 0 })
    }
}

// Every `fmt_{node}` function writes an indent
// at the beginning.
// It does _not_ write a newline at the end.

fn fmt_expr(expr: &Expr, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    match expr {
        Expr::Ind(e) => fmt_ind(e, f, indent),
        Expr::Vcon(e) => fmt_vcon(e, f, indent),
        Expr::Match(e) => fmt_match(e, f, indent),
        Expr::Fun(e) => fmt_fun(e, f, indent),
        Expr::App(e) => fmt_app(e, f, indent),
        Expr::For(e) => fmt_for(e, f, indent),
        Expr::Deb(e) => fmt_deb(e, f, indent),
        Expr::Universe(e) => fmt_universe(e, f, indent),
    }
}

fn fmt_ind(ind: &RcSemHashed<Ind>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    let universe_level = ind.value.universe_level.0;
    write!(f, "{indent}(\n{i1}ind\n{i1}Type{universe_level}\n")?;
    fmt_str_literal(ind.value.name.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_expressions(ind.value.index_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_vcon_defs(ind.value.vcon_defs.clone(), f, i1)?;
    write!(f, "\n{indent})")?;

    Ok(())
}

fn fmt_str_literal(
    str_literal: Rc<StringValue>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}

fn fmt_parenthesized_vcon_defs(
    parenthesized_vcon_defs: RcSemHashed<Box<[VconDef]>>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}

fn fmt_vcon(vcon: &RcSemHashed<Vcon>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_match(m: &RcSemHashed<Match>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_fun(fun: &RcSemHashed<Fun>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_app(app: &RcSemHashed<App>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_for(for_: &RcSemHashed<For>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_deb(deb: &RcSemHashed<DebNode>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let index = deb.value.deb.0;
    write!(f, "{indent}Type{index}")
}

fn fmt_universe(
    universe: &RcSemHashed<UniverseNode>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    let level = universe.value.level.0;
    write!(f, "{indent}Type{level}")
}

fn fmt_parenthesized_expressions(
    parenthesized_expressions: RcSemHashed<Box<[Expr]>>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}
