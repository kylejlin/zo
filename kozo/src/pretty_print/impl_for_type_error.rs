use super::*;

use crate::{
    syntax_tree::{ipist::rc_hashed, ipist_to_ast::*},
    typecheck::TypeError,
};

impl Display for PrettyPrint<'_, TypeError> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match &self.0 {
            TypeError::InvalidDeb { deb, tcon_len } => f
                .debug_struct("TypeError::InvalidDeb")
                .field("deb", &deb.value)
                .field("tcon_len", tcon_len)
                .finish(),

            TypeError::InvalidVconIndex(vcon) => {
                let mut converter = IpistToAstConverter::default();
                let vcon_ast = converter.convert_vcon(rc_hashed(vcon.clone()));
                f.debug_struct("TypeError::InvalidVconIndex")
                    .field("vcon", &vcon_ast.pretty_printed())
                    .finish()
            }

            TypeError::UnexpectedNonTypeExpression { expr, type_ } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::UnexpectedNonTypeExpression")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("type_", &type_.raw().pretty_printed())
                    .finish()
            }

            TypeError::UniverseInconsistencyInIndDef {
                index_or_param_type,
                level,
                ind,
            } => {
                let mut converter = IpistToAstConverter::default();
                let index_or_param_type_ast = converter.convert(index_or_param_type.clone());
                let ind_ast = converter.convert_ind(rc_hashed(ind.clone()));
                f.debug_struct("TypeError::UniverseInconsistencyInIndDef")
                    .field(
                        "index_or_param_type",
                        &index_or_param_type_ast.pretty_printed(),
                    )
                    .field("level", &level)
                    .field("ind", &ind_ast.pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfIndexArguments {
                def,
                expected,
                actual,
            } => {
                let mut converter = IpistToAstConverter::default();
                let def_ast = converter.convert_vcon_def(def.clone());
                f.debug_struct("TypeError::WrongNumberOfIndexArguments")
                    .field("def", &def_ast.pretty_printed())
                    .field("expected", expected)
                    .field("actual", actual)
                    .finish()
            }

            TypeError::NonInductiveMatcheeType { expr, type_ } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::NonInductiveMatcheeType")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("type_", &type_.raw().pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfMatchCases {
                match_,
                matchee_type_ind,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::WrongNumberOfMatchCases")
                    .field("match_", &match_ast.pretty_printed())
                    .field(
                        "matchee_type_ind",
                        &rc_hashed(matchee_type_ind.raw().clone()).pretty_printed(),
                    )
                    .finish()
            }

            TypeError::WrongMatchCaseArity {
                actual_node,
                expected,
                match_,
                match_case_index,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::WrongMatchCaseArity")
                    .field("actual_node", &actual_node.value)
                    .field("expected", expected)
                    .field("match_", &match_ast.pretty_printed())
                    .field("match_case_index", match_case_index)
                    .finish()
            }

            TypeError::IllegallyDismissedMatchCase {
                match_,
                match_case_index,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::IllegallyDismissedMatchCase")
                    .field("match_", &match_ast.pretty_printed())
                    .field("match_case_index", match_case_index)
                    .finish()
            }

            TypeError::TypeMismatch {
                expr,
                expected_type,
                actual_type,
            } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::TypeMismatch")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("expected_type", &expected_type.raw().pretty_printed())
                    .field("actual_type", &actual_type.raw().pretty_printed())
                    .finish()
            }

            TypeError::CalleeTypeIsNotAForExpression { app, callee_type } => {
                let mut converter = IpistToAstConverter::default();
                let app_ast = converter.convert_app(rc_hashed(app.clone()));
                f.debug_struct("TypeError::CalleeTypeIsNotAForExpression")
                    .field("app", &app_ast.pretty_printed())
                    .field("callee_type", &callee_type.raw().pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfAppArguments {
                app,
                callee_type,
                expected,
                actual,
            } => {
                let mut converter = IpistToAstConverter::default();
                let app_ast = converter.convert_app(rc_hashed(app.clone()));
                f.debug_struct("TypeError::WrongNumberOfAppArguments")
                    .field("app", &app_ast.pretty_printed())
                    .field(
                        "callee_type",
                        &rc_hashed(callee_type.raw().clone()).pretty_printed(),
                    )
                    .field("expected", expected)
                    .field("actual", actual)
                    .finish()
            }
        }
    }
}
