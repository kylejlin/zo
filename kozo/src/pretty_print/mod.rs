use std::fmt::{Display, Formatter, Result as FmtResult};

pub const SOFT_TAB_SIZE: usize = 4;

#[derive(Clone, Debug, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PrettyPrinted<'a, T>(pub &'a T);

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct SoftTabCount(pub usize);

fn get_indent_str(count: SoftTabCount) -> String {
    " ".repeat(count.0 * SOFT_TAB_SIZE)
}

mod impl_for_ast;
