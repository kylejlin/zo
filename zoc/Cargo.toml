[package]
name = "zoc"
version = "0.1.0"
edition = "2021"

[dependencies]
rustc-hash = "1.1.0"

[dev-dependencies]
pretty_assertions = "1.3.0"
insta = "1.29.0"

[build-dependencies]
kiki = "7"
walkdir = "2"
sha256 = "1.1.4"
