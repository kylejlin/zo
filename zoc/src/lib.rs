pub mod check_erasability;
pub mod eval;
pub mod hash;
pub mod pretty_print;
pub mod syntax_tree;
pub mod typecheck;

#[cfg(test)]
pub mod test_utils;
