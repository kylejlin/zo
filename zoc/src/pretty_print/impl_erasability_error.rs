use super::*;

use crate::check_erasability::ErasabilityError;

impl Display for PrettyPrint<'_, ErasabilityError> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match &self.0 {
            ErasabilityError::MatcheeTypeTypeIsErasableButReturnTypeTypeIsNotErasable {
                match_,
                matchee_type_type,
                match_return_type_type,
            } => f
                .debug_struct(
                    "ErasabilityError::MatcheeTypeTypeIsErasableButReturnTypeTypeIsNotErasable",
                )
                .field("match_", &match_.pretty_printed())
                .field("matchee_type_type", &matchee_type_type.pretty_printed())
                .field(
                    "match_return_type_type",
                    &match_return_type_type.pretty_printed(),
                )
                .finish(),
        }
    }
}
