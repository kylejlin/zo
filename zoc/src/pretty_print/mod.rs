use std::{
    fmt::{Debug, Display, Formatter, Result as FmtResult},
    rc::Rc,
};

#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PrettyPrint<'a, T>(pub &'a T);

impl<'a, T> Debug for PrettyPrint<'a, T>
where
    PrettyPrint<'a, T>: Display,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        Display::fmt(self, f)
    }
}

mod impl_for_ast;
mod impl_for_type_error;

pub trait PrettyPrinted {
    fn pretty_printed(&self) -> PrettyPrint<'_, Self>
    where
        Self: Sized;
}

impl<T> PrettyPrinted for T {
    fn pretty_printed(&self) -> PrettyPrint<'_, Self>
    where
        Self: Sized,
    {
        PrettyPrint(self)
    }
}

pub trait PrettyUnwrap {
    type Output;

    fn pretty_unwrap(self) -> Self::Output;
}

pub trait PrettyUnwrapErr {
    type Output;

    fn pretty_unwrap_err(self) -> Self::Output;
}

impl<T, E> PrettyUnwrap for Result<T, E>
where
    for<'a> PrettyPrint<'a, E>: Display,
{
    type Output = T;

    fn pretty_unwrap(self) -> T {
        match self {
            Ok(v) => v,
            Err(e) => {
                // We use the "#" flag so that if the
                // Display impl of `PrettyPrint<E>` uses
                // `Formatter:debug_struct` (or `debug_tuple`, etc.),
                // the output will be printed with easy-to-read
                // indentation.
                panic!(
                    "called `Result::unwrap()` on an `Err` value:\n{:#}",
                    PrettyPrint(&e)
                );
            }
        }
    }
}

impl<T, E> PrettyUnwrapErr for Result<T, E>
where
    for<'a> PrettyPrint<'a, T>: Display,
{
    type Output = E;

    fn pretty_unwrap_err(self) -> E {
        match self {
            Ok(v) => {
                panic!(
                    "called `Result::unwrap_err()` on an `Ok` value:\n{:#}",
                    PrettyPrint(&v)
                );
            }
            Err(e) => e,
        }
    }
}

pub const SOFT_TAB: &str = "    ";

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct Indentation {
    pub soft_tab_count: usize,
}

impl Indentation {
    pub fn incremented(self) -> Self {
        Self {
            soft_tab_count: self.soft_tab_count + 1,
        }
    }
}

impl Display for Indentation {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        for _ in 0..self.soft_tab_count {
            f.write_str(SOFT_TAB)?;
        }
        Ok(())
    }
}
