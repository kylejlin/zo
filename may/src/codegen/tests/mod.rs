use super::*;

use zoc::pretty_print::PrettyPrint;

fn parse_or_panic(src: &str) -> mnode::Expr {
    let tokens = crate::lexer::lex(src).unwrap();
    crate::parser::parse(tokens).unwrap()
}

fn assert_expr_is_well_typed_under_empty_tcon(ast: znode::Expr) {
    use zoc::{
        eval::Normalized,
        pretty_print::PrettyUnwrap,
        syntax_tree::{lexer::lex as zo_lex, parser::parse as zo_parse},
        typecheck::{LazyTypeContext, TypeChecker},
    };

    let src = PrettyPrint(&ast).to_string();
    let tokens = zo_lex(&src).unwrap();
    let ost = zo_parse(tokens).unwrap();

    let empty = Normalized::<[_; 0]>::new();
    TypeChecker::default()
        .get_type(
            ost.into(),
            LazyTypeContext::Base(empty.as_ref().convert_ref()),
        )
        .pretty_unwrap();
}

fn assert_expr_is_ill_typed_under_empty_tcon(ast: znode::Expr) {
    use zoc::{
        eval::Normalized,
        pretty_print::PrettyUnwrapErr,
        syntax_tree::{lexer::lex as zo_lex, parser::parse as zo_parse},
        typecheck::{LazyTypeContext, TypeChecker},
    };

    let src = PrettyPrint(&ast).to_string();
    let tokens = zo_lex(&src).unwrap();
    let ost = zo_parse(tokens).unwrap();

    let empty = Normalized::<[_; 0]>::new();
    TypeChecker::default()
        .get_type(
            ost.into(),
            LazyTypeContext::Base(empty.as_ref().convert_ref()),
        )
        .map(Normalized::into_raw)
        .pretty_unwrap_err();
}

fn assert_final_expression_and_topright_defs_are_well_typed_under_empty_tcon(
    src: &str,
) -> znode::Expr {
    let cst = parse_or_panic(src);
    let (converted_leaf, topright_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in topright_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    converted_leaf
}

#[test]
fn ill_typed_unused_def_does_not_affect_converted_leaf() {
    let src = r#"
ind Nat
    case zero
    case succ(_: Nat)
    return Set0

fun bad(n: Nat): Nat
    match n
    case zero:
        zero
    // missing `succ` case
    return1 Nat

succ(succ(zero))
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    // `substitutable_defs` should be `[Nat, zero, succ, bad]`.
    assert_eq!(4, substitutable_defs.len());

    // `Nat`, `zero`, and `succ` should be well typed.
    assert_expr_is_well_typed_under_empty_tcon(substitutable_defs[0].clone());
    assert_expr_is_well_typed_under_empty_tcon(substitutable_defs[1].clone());
    assert_expr_is_well_typed_under_empty_tcon(substitutable_defs[2].clone());

    // `bad` should be ill typed.
    assert_expr_is_ill_typed_under_empty_tcon(substitutable_defs[3].clone());

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn two() {
    let src = include_str!("samples/two.may");
    let converted_leaf =
        assert_final_expression_and_topright_defs_are_well_typed_under_empty_tcon(src);
    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn add_two_three() {
    let src = r#"
ind Nat
    case zero
    case succ(_: Nat)
    return Set0

let _2 = succ(succ(zero))

let _3 = succ(_2)

fun add(-a: Nat, b: Nat): Nat
    match a
    case zero:
        b
    case succ(a_pred):
        succ(add(a_pred, b))
    return1 Nat

add(_2, _3)
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn list() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
List
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn list_nil() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
nil
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn list_cons() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
cons
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn rev() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0

fun rev(T: Set0, xs: List(T)): List(T)
    fun helper(-xs: List(T), acc: List(T)): List(T)
        match xs
        case nil:
            acc
        case cons(x, xs):
            helper(xs, cons(T)(x, acc))
        return1 List(T)
    helper(xs, nil(T))

ind Abc
    case a
    case b
    case c
    return Set0

let acons = cons(Abc)
let anil = nil(Abc)

let a_b_c = acons(a, acons(b, acons(c, anil)))

rev(Abc, a_b_c)
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn eq() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0
Eq
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn eq_refl() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0
refl
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn eq_commutative() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun(T: Set0, a: T, b: T, eq: Eq(T, a)(b)): Eq(T, b)(a)
    match eq
    case refl:
        refl(T, a)
    use [c] return Eq(T, c)(a)
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn eq_transitive() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun(
    T: Set0,
    a: T,
    b: T,
    c: T,
    ab: Eq(T, a)(b),
    bc: Eq(T, b)(c),
): Eq(T, a)(c)
    let f =
        match ab
        case refl:
            afun(ac: Eq(T, a)(c)): Eq(T, a)(c)
                ac
        use [ina_outb]
        return For(_: Eq(T, ina_outb)(c)) -> Eq(T, a)(c)
    f(bc)
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn add_n_zero() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

ind Nat
    case zero
    case succ(_: Nat)
    return Set0

fun add(-a: Nat, b: Nat): Nat
    match a
    case zero:
        b
    case succ(a_pred):
        succ(add(a_pred, b))
    return1 Nat

fun add_n_zero(-n: Nat): Eq(Nat, n)(add(n, zero))
    match n
    case zero:
        refl(Nat, zero)
    case succ(n_pred):
        match add_n_zero(n_pred)
        case refl:
            refl(Nat, succ(n_pred))
        use [in_n_pred_out_add_n_pred_zero]
        return Eq(Nat, succ(n_pred))(succ(in_n_pred_out_add_n_pred_zero))
    use n_capp
    return1 Eq(Nat, n_capp)(add(n_capp, zero))

add_n_zero
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn add_n_succ_m() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

ind Nat
    case zero
    case succ(_: Nat)
    return Set0

fun add(-a: Nat, b: Nat): Nat
    match a
    case zero:
        b
    case succ(a_pred):
        succ(add(a_pred, b))
    return1 Nat

fun add_n_succ_m(-n: Nat, m: Nat): Eq(Nat, succ(add(n, m)))(add(n, succ(m)))
    match n
    case zero:
        refl(Nat, succ(m))
    case succ(n_pred):
        // Goal: Eq(Nat, succ(succ(add(n_pred, m))))(succ(add(n_pred, succ(m))))
        match add_n_succ_m(n_pred, m)
        case refl:
            refl(Nat, succ(succ(add(n_pred, m))))
        use [in_succ_add_out_add_succ]
        return Eq(Nat, succ(succ(add(n_pred, m))))(succ(in_succ_add_out_add_succ))
    use n_capp
    return1 Eq(Nat, succ(add(n_capp, m)))(add(n_capp, succ(m)))

add_n_succ_m
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn add_commutative() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

ind Nat
    case zero
    case succ(_: Nat)
    return Set0

fun add(-a: Nat, b: Nat): Nat
    match a
    case zero:
        b
    case succ(a_pred):
        succ(add(a_pred, b))
    return1 Nat

fun add_n_zero(-n: Nat): Eq(Nat, n)(add(n, zero))
    match n
    case zero:
        refl(Nat, zero)
    case succ(n_pred):
        match add_n_zero(n_pred)
        case refl:
            refl(Nat, succ(n_pred))
        use [in_n_pred_out_add_n_pred_zero]
        return Eq(Nat, succ(n_pred))(succ(in_n_pred_out_add_n_pred_zero))
    use n_capp
    return1 Eq(Nat, n_capp)(add(n_capp, zero))

fun add_n_succ_m(-n: Nat, m: Nat): Eq(Nat, succ(add(n, m)))(add(n, succ(m)))
    match n
    case zero:
        refl(Nat, succ(m))
    case succ(n_pred):
        // Goal: Eq(Nat, succ(succ(add(n_pred, m))))(succ(add(n_pred, succ(m))))
        match add_n_succ_m(n_pred, m)
        case refl:
            refl(Nat, succ(succ(add(n_pred, m))))
        use [in_succ_add_out_add_succ]
        return Eq(Nat, succ(succ(add(n_pred, m))))(succ(in_succ_add_out_add_succ))
    use n_capp
    return1 Eq(Nat, succ(add(n_capp, m)))(add(n_capp, succ(m)))

afun add_commutative(-n: Nat, m: Nat): Eq(Nat, add(n, m))(add(m, n))
    match n
    case zero:
        match add_n_zero(m)
        case refl:
            refl(Nat, m)
        use [in_m_out_add_m_zero]
        return Eq(Nat, m)(in_m_out_add_m_zero)
    case succ(n_pred):
        // Goal: Eq(Nat, succ(add(n_pred, m)))(add(m, succ(n_pred)))
        match add_n_succ_m(m, n_pred)
        case refl:
            // Goal: Eq(Nat, succ(add(n_pred, m)))(succ(add(m, n_pred)))
            match add_commutative(n_pred, m)
            case refl:
                refl(Nat, succ(add(n_pred, m)))
            use [in_npred_m_out_m_apred]
            return Eq(Nat, succ(add(n_pred, m)))(succ(in_npred_m_out_m_apred))
        use [in_succ_add_out_add_succ]
        return Eq(Nat, succ(add(n_pred, m)))(in_succ_add_out_add_succ)
    use n_capp
    return1 Eq(Nat, add(n_capp, m))(add(m, n_capp))
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn eq_implies_substitutable() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun eq_implies_substitutable(
    T: Set0,
    a: T,
    b: T,
    eq: Eq(T, a)(b),
    P: For(_: T) -> Prop0,
    pa: P(a),
): P(b)
    match eq
    case refl:
        pa
    use [in_a_out_b]
    return P(in_a_out_b)
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn substitutable_implies_eq() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun substitutable_implies_eq(
    T: Set0,
    a: T,
    b: T,
    sub: For(P: For(_: T) -> Prop0, pa: P(a)) -> P(b),
): Eq(T, a)(b)
    fun eq_a(c: T): Prop0
        Eq(T, a)(c)
    sub(eq_a, refl(T, a))
"#;
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}

#[test]
fn stdlib() {
    let src = include_str!("samples/std.may");
    let cst = parse_or_panic(src);
    let (converted_leaf, substitutable_defs) = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(converted_leaf.clone());

    for def in substitutable_defs {
        assert_expr_is_well_typed_under_empty_tcon(def);
    }

    insta::assert_display_snapshot!(PrettyPrint(&converted_leaf));
}
