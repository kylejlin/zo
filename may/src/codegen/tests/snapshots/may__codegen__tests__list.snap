---
source: may/src/codegen/tests/mod.rs
expression: PrettyPrint(&converted_leaf)
---
(
    fun
    nonrec
    (
        Set0
    )
    Set0
    (
        ind
        Set0
        "List"
        ()
        (
            (
                (
                    2
                    1
                )
                ()
            )
            (
                ()
                ()
            )
        )
    )
)
